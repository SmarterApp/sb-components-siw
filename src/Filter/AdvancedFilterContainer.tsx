import * as React from "react";
import { AdvancedFilter } from "./AdvancedFilter";
import { SelectedFilterIndicator } from "./SelectedFilterIndicator";
import {
  advFilterSelect,
  AdvancedFilterCategoryModel,
  AdvancedFiltersModel,
  FilterOptionModel,
  OptionTypeModel,
  FilterType,
  FilterCategoryModel
} from "./FilterModels";

/**
 * AdvancedFilterContainer props
 * @interface AdvancedFilterContainerProps
 * @member {AdvancedFilterCategoryModel[]} filterOptions
 * @method {(selected: AdvancedFilterCategoryModel[]) => void} onUpdateFilterOptions
 * @member {boolean} isNested
 * @member {string?} pageTitle
 * @member {string?} filterId
 */
export interface AdvancedFilterContainerProps {
  filterCategories: AdvancedFilterCategoryModel[];
  onUpdateFilter: (
    selected?: AdvancedFilterCategoryModel[],
    changed?: FilterType
  ) => void;
  isNested?: boolean;
  pageTitle?: string;
  filterId?: string;
}
/**
 * AdvancedFilterContainer state
 * @interface AdvancedFilterContainerState
 * @member {boolean} expanded
 */
export interface AdvancedFilterContainerState {
  expanded: boolean;
}

/**
 * The AdvancedFilterContainer is a collapsible menu that displays AdvancedFilters
 * that, when clicked, calls `this.props.onClick()`
 * @class AdvancedFilterContainer
 * @extends {React.Component<AdvancedFilterContainerProps, AdvancedFilterContainerState>}
 */
export class AdvancedFilterContainer extends React.Component<
  AdvancedFilterContainerProps,
  AdvancedFilterContainerState
> {
  constructor(props: AdvancedFilterContainerProps) {
    super(props);
    this.state = {
      expanded: this.props.isNested ? true : false
    };
  }

  handleClick = () => {
    this.setState({ expanded: !this.state.expanded });
  };

  handleFilterSelect(
    category: AdvancedFilterCategoryModel,
    option?: FilterOptionModel
  ) {
    const { onUpdateFilter, filterCategories } = this.props;
    const newFilters = advFilterSelect(filterCategories, category, option);
    onUpdateFilter(newFilters, category.code);
  }

  /**
   * Resets each of the filter options for each category.
   */
  resetFilters() {
    const { filterCategories } = this.props;
    filterCategories.forEach(cat =>
      cat.filterOptions.forEach(fo => (fo.isSelected = false))
    );
    this.props.onUpdateFilter(filterCategories);
  }

  /**
   * Returns true if one or more filter options are selected in any filter category.
   * False if otherwise
   */
  hasActiveFilterIndicators() {
    const { filterCategories } = this.props;
    let active = false;
    filterCategories.forEach(cat => {
      if (!cat.disabled) {
        cat.filterOptions.forEach(opt => {
          if (opt.isSelected) {
            active = true;
          }
        });
      }
    });

    return active;
  }

  /**
   * Builds and returns a list of JSX.Elements that shows which filter
   * options are currently selected
   */
  renderSelectedFilterIndicators() {
    const tags: JSX.Element[] = [];
    this.props.filterCategories.forEach(cat => {
      if (!cat.disabled && cat.optionType !== OptionTypeModel.inputBox) {
        cat.filterOptions.forEach(opt => {
          if (opt.key !== "0" && opt.isSelected) {
            tags.push(
              <SelectedFilterIndicator
                key={cat.label + opt.key}
                category={cat}
                option={opt}
                onClick={() => this.handleFilterSelect(cat, opt)}
              />
            );
          }
        });
      }
    });

    return <div className="filter-status">{tags}</div>;
  }

  /**
   * Renders the array of filter categories and their respective filter options.
   */
  renderFilterCategories() {
    const { filterCategories } = this.props;
    const filterCats = filterCategories.map((category, i) => {
      if (category.show == true) {
        return (
          <AdvancedFilter
            key={i}
            {...category}
            onFilterOptionSelect={opt => this.handleFilterSelect(category, opt)}
          />
        );
      }
    });

    return (
      <div className="filter-body" aria-live="polite" aria-relevant="additions">
        {filterCats}
      </div>
    );
  }

  /**
   * Renders the reset button
   */
  renderResetButton(): JSX.Element | undefined {
    let content: JSX.Element | undefined;
    if (this.hasActiveFilterIndicators() && !this.props.isNested) {
      content = (
        <button
          onClick={() => this.resetFilters()}
          className="btn btn-default filter-reset-btn"
        >
          Reset Filters
        </button>
      );
    }

    return content;
  }

  /**
   * Renders the button that, when clicked, expands or collapses the advanced filter.
   */
  renderExpandButton() {
    const { expanded } = this.state;
    const className = expanded ? "fa fa-chevron-down" : "fa fa-chevron-right";
    const buttonText = expanded ? "Collapse " : "Expand ";

    return (
      <button
        onClick={this.handleClick}
        className="btn btn-default filter-expand-btn"
      >
        {buttonText}
        <span className={className} />
      </button>
    );
  }

  /**
   * Renders the page title.
   */
  renderPageTitle(): JSX.Element | undefined {
    let content: JSX.Element | undefined;
    if (this.props.pageTitle) {
      content = (
        <h1>
          <span className="filter-page-title">{this.props.pageTitle}</span>
        </h1>
      );
    }

    return content;
  }
  /**
   * Renders the portion of the Advanced filter container that will always be visible
   * and dictates expansion of the filter menu, essentially the 'header' of the component.
   */
  renderCollapsedFilterContainer(): JSX.Element {
    return (
      <div className="adv-filter-group">
        {this.renderPageTitle()}
        <div className="adv-filter-header">
          <div className="adv-filter-title">
            <h3>
              <span className="fa fa-tasks" />&nbsp;Advanced Filters
            </h3>
          </div>
          <div className="adv-control-btns">
            {this.renderResetButton()}
            {this.renderExpandButton()}
          </div>
        </div>
        {this.renderSelectedFilterIndicators()}
      </div>
    );
  }

  /**
   * Render the expanded filter container
   */
  renderExpanded(): JSX.Element | undefined {
    let content: JSX.Element | undefined;
    if (this.state.expanded) {
      content = (
        <div className="adv-filter-container-expanded">
          {this.renderFilterCategories()}
        </div>
      );
    }

    return content;
  }

  render() {
    const id = this.props.filterId ? this.props.filterId : "";

    return (
      <div id={id} className="section section-light adv-filter-container">
        {this.renderCollapsedFilterContainer()}
        {this.renderExpanded()}
      </div>
    );
  }
}
