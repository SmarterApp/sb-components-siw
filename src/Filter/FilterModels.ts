import {
  SubjectModel,
  SearchAPIParamsModel,
  GradeLevels,
  GradeLevel
} from "@src/index";

/**
 * Updates a category with the filter option that was selected
 * @param {AdvancedFilterCategoryModel[]} categories
 * @param {AdvancedFilterCategoryModel} selectedCat
 * @param {FilterOptionModel} [option]
 */
export const advFilterSelect = (
  categories: AdvancedFilterCategoryModel[],
  selectedCat: AdvancedFilterCategoryModel,
  option?: FilterOptionModel
): AdvancedFilterCategoryModel[] | undefined => {
  const allPressed = option === undefined && selectedCat.displayAllButton;
  if (!selectedCat.disabled) {
    const categoryIndex = categories.indexOf(selectedCat);

    categories[categoryIndex] = advFilterCategorySelect(
      categories[categoryIndex],
      option
    );

    return categories;
  }
};

export function advFilterCategorySelect(
  selectedCat: AdvancedFilterCategoryModel,
  option?: FilterOptionModel
): AdvancedFilterCategoryModel {
  let newCategory = { ...selectedCat };
  const options = newCategory.filterOptions.slice();
  const allPressed =
    (option === undefined || option.filterType === undefined) &&
    newCategory.displayAllButton;

  if (option != undefined && option.filterType === "SearchItemId") {
    newCategory.filterOptions[0] = option;
    return newCategory;
  }

  if (!newCategory.disabled) {
    if (allPressed) {
      options.forEach((opt: FilterOptionModel) => (opt.isSelected = false));
    } else if (option) {
      const optionIdx = options.indexOf(option);
      options[optionIdx].isSelected = !option.isSelected;
      if (!selectedCat.isMultiSelect) {
        options.forEach((opt: FilterOptionModel) => {
          opt.isSelected = opt === option ? opt.isSelected : false;
        });
      }
    }
  }

  return newCategory;
}

export enum OptionTypeModel {
  inputBox,
  button,
  radioBtn,
  DropDown,
  AdvFilter
}

export enum FilterType {
  Subject = "Subject",
  Grade = "Grade",
  Claim = "Claim",
  Performance = "Performance",
  Target = "Target",
  CAT = "CAT",
  InteractionType = "InteractionType",
  Calculator = "Calculator",
  TechnologyType = "TechnologyType", // Contains Performance and CAT
  SearchItemId = "SearchItemId",
  TestNames = "TestNames",
  CoreStandards = "CoreStandards",
  ReleaseDate = "ReleaseDate"
}

export interface FilterOptionModel {
  label: string;
  key: string;
  isSelected: boolean;
  filterType?: FilterType;
}

export interface BasicFilterCategoryModel extends FilterCategoryModel {
  optionType: OptionTypeModel;
  placeholderText?: string;
  hideSelectMessage?: boolean;
  displayAllButton?: boolean;
}

export interface FilterCategoryModel {
  disabled: boolean;
  label: string;
  filterOptions: FilterOptionModel[];
  helpText?: string;
  emptyOptionsText?: string;
  code: FilterType;
  isMultiSelect?: boolean;
  show?: boolean;
}

export interface AdvancedFilterCategoryModel extends FilterCategoryModel {
  displayAllButton: boolean;
  optionType?: OptionTypeModel;
  placeholderText?: string;
  hideSelectMessage?: boolean;
}

// tslint:disable-next-line: no-empty-interface
export interface TechType extends SubjectModel {}

export interface AdvancedFiltersModel {
  subjects: AdvancedFilterCategoryModel;
  grades: AdvancedFilterCategoryModel;
  techTypes: AdvancedFilterCategoryModel;
}
