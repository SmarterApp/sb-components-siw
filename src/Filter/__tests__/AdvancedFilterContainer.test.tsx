import * as React from "react";
import { AdvancedFilterContainer } from "../AdvancedFilterContainer";
import { AdvancedFilterCategoryModel } from "../FilterModels";
import {
  mockAdvancedFilterCategoriesAll,
  mockAdvancedFilterCategoriesSelected
} from "../../../mocks/Filter/mocks";
import { shallow } from "enzyme";

describe("AdvancedFilterContainer", () => {
  const updateFilter = jest.fn(
    (filterCategories: AdvancedFilterCategoryModel[]) => {
      wrapper.setProps({ filterCategories });
    }
  );

  const wrapper = shallow(
    <AdvancedFilterContainer
      pageTitle="Advanced Filter"
      onUpdateFilter={updateFilter}
      filterCategories={mockAdvancedFilterCategoriesAll}
    />
  );

  const wrapper1 = shallow(
    <AdvancedFilterContainer
      onUpdateFilter={updateFilter}
      filterCategories={mockAdvancedFilterCategoriesSelected}
    />
  );

  it("expands and collapses on click", () => {
    wrapper.find(".filter-expand-btn").simulate("click");
    expect(wrapper).toMatchSnapshot();
    wrapper.find(".filter-expand-btn").simulate("click");
    expect(wrapper).toMatchSnapshot();
  });

  it("'Reset Filters' button resets filters", () => {
    wrapper1
      .findWhere(node => node.type() === "button")
      .at(0)
      .simulate("click");
    expect(wrapper1).toMatchSnapshot();
    wrapper1.find(".filter-reset-btn").simulate("click");
    expect(wrapper1).toMatchSnapshot();
  });
});
