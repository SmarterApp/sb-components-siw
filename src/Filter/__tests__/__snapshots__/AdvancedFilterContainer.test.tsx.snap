// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedFilterContainer 'Reset Filters' button resets filters 1`] = `
<div
  className="section section-light adv-filter-container"
  id=""
>
  <div
    className="adv-filter-group"
  >
    <div
      className="adv-filter-header"
    >
      <div
        className="adv-filter-title"
      >
        <h3>
          <span
            className="fa fa-tasks"
          />
           Advanced Filters
        </h3>
      </div>
      <div
        className="adv-control-btns"
      >
        <button
          className="btn btn-default filter-reset-btn"
          onClick={[Function]}
        >
          Reset Filters
        </button>
        <button
          className="btn btn-default filter-expand-btn"
          onClick={[Function]}
        >
          Expand 
          <span
            className="fa fa-chevron-right"
          />
        </button>
      </div>
    </div>
    <div
      className="filter-status"
    >
      <Component
        category={
          Object {
            "code": "Grade",
            "disabled": false,
            "displayAllButton": true,
            "filterOptions": Array [
              Object {
                "filterType": "Subject",
                "isSelected": false,
                "key": "ELA",
                "label": "English Literacy & Arts",
              },
              Object {
                "filterType": "Subject",
                "isSelected": false,
                "key": "Math",
                "label": "Maths",
              },
            ],
            "helpText": "Grade Help",
            "isMultiSelect": false,
            "label": "Grade",
            "show": true,
          }
        }
        key="GradeELA"
        onClick={[Function]}
        option={
          Object {
            "filterType": "Subject",
            "isSelected": false,
            "key": "ELA",
            "label": "English Literacy & Arts",
          }
        }
      />
      <Component
        category={
          Object {
            "code": "Grade",
            "disabled": false,
            "displayAllButton": true,
            "filterOptions": Array [
              Object {
                "filterType": "Subject",
                "isSelected": false,
                "key": "ELA",
                "label": "English Literacy & Arts",
              },
              Object {
                "filterType": "Subject",
                "isSelected": false,
                "key": "Math",
                "label": "Maths",
              },
            ],
            "helpText": "Grade Help",
            "isMultiSelect": false,
            "label": "Grade",
            "show": true,
          }
        }
        key="GradeMath"
        onClick={[Function]}
        option={
          Object {
            "filterType": "Subject",
            "isSelected": false,
            "key": "Math",
            "label": "Maths",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`AdvancedFilterContainer 'Reset Filters' button resets filters 2`] = `
<div
  className="section section-light adv-filter-container"
  id=""
>
  <div
    className="adv-filter-group"
  >
    <div
      className="adv-filter-header"
    >
      <div
        className="adv-filter-title"
      >
        <h3>
          <span
            className="fa fa-tasks"
          />
           Advanced Filters
        </h3>
      </div>
      <div
        className="adv-control-btns"
      >
        <button
          className="btn btn-default filter-reset-btn"
          onClick={[Function]}
        >
          Reset Filters
        </button>
        <button
          className="btn btn-default filter-expand-btn"
          onClick={[Function]}
        >
          Expand 
          <span
            className="fa fa-chevron-right"
          />
        </button>
      </div>
    </div>
    <div
      className="filter-status"
    >
      <Component
        category={
          Object {
            "code": "Grade",
            "disabled": false,
            "displayAllButton": true,
            "filterOptions": Array [
              Object {
                "filterType": "Subject",
                "isSelected": false,
                "key": "ELA",
                "label": "English Literacy & Arts",
              },
              Object {
                "filterType": "Subject",
                "isSelected": false,
                "key": "Math",
                "label": "Maths",
              },
            ],
            "helpText": "Grade Help",
            "isMultiSelect": false,
            "label": "Grade",
            "show": true,
          }
        }
        key="GradeELA"
        onClick={[Function]}
        option={
          Object {
            "filterType": "Subject",
            "isSelected": false,
            "key": "ELA",
            "label": "English Literacy & Arts",
          }
        }
      />
      <Component
        category={
          Object {
            "code": "Grade",
            "disabled": false,
            "displayAllButton": true,
            "filterOptions": Array [
              Object {
                "filterType": "Subject",
                "isSelected": false,
                "key": "ELA",
                "label": "English Literacy & Arts",
              },
              Object {
                "filterType": "Subject",
                "isSelected": false,
                "key": "Math",
                "label": "Maths",
              },
            ],
            "helpText": "Grade Help",
            "isMultiSelect": false,
            "label": "Grade",
            "show": true,
          }
        }
        key="GradeMath"
        onClick={[Function]}
        option={
          Object {
            "filterType": "Subject",
            "isSelected": false,
            "key": "Math",
            "label": "Maths",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`AdvancedFilterContainer expands and collapses on click 1`] = `
<div
  className="section section-light adv-filter-container"
  id=""
>
  <div
    className="adv-filter-group"
  >
    <h1>
      <span
        className="filter-page-title"
      >
        Advanced Filter
      </span>
    </h1>
    <div
      className="adv-filter-header"
    >
      <div
        className="adv-filter-title"
      >
        <h3>
          <span
            className="fa fa-tasks"
          />
           Advanced Filters
        </h3>
      </div>
      <div
        className="adv-control-btns"
      >
        <button
          className="btn btn-default filter-expand-btn"
          onClick={[Function]}
        >
          Collapse 
          <span
            className="fa fa-chevron-down"
          />
        </button>
      </div>
    </div>
    <div
      className="filter-status"
    />
  </div>
  <div
    className="adv-filter-container-expanded"
  >
    <div
      aria-live="polite"
      aria-relevant="additions"
      className="filter-body"
    >
      <AdvancedFilter
        code="Grade"
        disabled={false}
        displayAllButton={true}
        filterOptions={
          Array [
            Object {
              "filterType": "Grade",
              "isSelected": false,
              "key": "7",
              "label": "Grades 3-5",
            },
            Object {
              "filterType": "Grade",
              "isSelected": false,
              "key": "56",
              "label": "Grades 6-8",
            },
            Object {
              "filterType": "Grade",
              "isSelected": false,
              "key": "960",
              "label": "High School",
            },
          ]
        }
        helpText="Grade Help"
        isMultiSelect={false}
        key="0"
        label="Grade"
        onFilterOptionSelect={[Function]}
        show={true}
      />
      <AdvancedFilter
        code="Subject"
        disabled={false}
        displayAllButton={false}
        filterOptions={
          Array [
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "EL&A",
              "label": "English Literacy & Arts",
            },
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "Mth",
              "label": "Maths",
            },
          ]
        }
        helpText="Subject Help"
        isMultiSelect={true}
        key="1"
        label="Multi"
        onFilterOptionSelect={[Function]}
        show={true}
      />
      <AdvancedFilter
        code="Subject"
        disabled={true}
        displayAllButton={false}
        filterOptions={
          Array [
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "Eng",
              "label": "English",
            },
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "Chem",
              "label": "Chemistry",
            },
          ]
        }
        helpText="Subject Help"
        isMultiSelect={false}
        key="2"
        label="Disabled"
        onFilterOptionSelect={[Function]}
        show={true}
      />
      <AdvancedFilter
        code="Subject"
        disabled={true}
        displayAllButton={false}
        filterOptions={
          Array [
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "Read",
              "label": "Reading",
            },
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "Hist",
              "label": "History",
            },
          ]
        }
        helpText="Subject Help"
        isMultiSelect={true}
        key="3"
        label="Disabled Multi"
        onFilterOptionSelect={[Function]}
        show={true}
      />
      <AdvancedFilter
        code="Subject"
        disabled={false}
        displayAllButton={true}
        filterOptions={
          Array [
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "USh",
              "label": "US History",
            },
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "Wh",
              "label": "World History",
            },
          ]
        }
        helpText="Subject Help"
        isMultiSelect={true}
        key="4"
        label="Multi All"
        onFilterOptionSelect={[Function]}
        show={true}
      />
      <AdvancedFilter
        code="Subject"
        disabled={false}
        displayAllButton={true}
        filterOptions={
          Array [
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "Calc",
              "label": "Calculus",
            },
            Object {
              "filterType": "Subject",
              "isSelected": false,
              "key": "Geom",
              "label": "Geometry",
            },
          ]
        }
        helpText="Subject Help"
        isMultiSelect={true}
        key="5"
        label="Multi All Disabled"
        onFilterOptionSelect={[Function]}
        show={true}
      />
    </div>
  </div>
</div>
`;

exports[`AdvancedFilterContainer expands and collapses on click 2`] = `
<div
  className="section section-light adv-filter-container"
  id=""
>
  <div
    className="adv-filter-group"
  >
    <h1>
      <span
        className="filter-page-title"
      >
        Advanced Filter
      </span>
    </h1>
    <div
      className="adv-filter-header"
    >
      <div
        className="adv-filter-title"
      >
        <h3>
          <span
            className="fa fa-tasks"
          />
           Advanced Filters
        </h3>
      </div>
      <div
        className="adv-control-btns"
      >
        <button
          className="btn btn-default filter-expand-btn"
          onClick={[Function]}
        >
          Expand 
          <span
            className="fa fa-chevron-right"
          />
        </button>
      </div>
    </div>
    <div
      className="filter-status"
    />
  </div>
</div>
`;
